File: badger/ui/run_panel.py
__all__ = ["RunPanel"]

import tkinter as tk
from collections.abc import Callable
from tkinter import ttk

from badger.ui.state import RunState


class RunPanel(ttk.LabelFrame):
    def __init__(
        self, parent: tk.Widget, *, on_run: Callable[[], None], on_stop: Callable[[], None]
    ) -> None:
        super().__init__(parent, text="Step 3  -  Run")

        row = ttk.Frame(self)
        row.pack(fill=tk.X, padx=6, pady=6)

        self.run_btn = ttk.Button(row, text="Start", command=on_run)
        self.run_btn.pack(side=tk.LEFT, padx=(0, 8))

        self.stop_btn = ttk.Button(row, text="Stop", command=on_stop)
        self.stop_btn.pack(side=tk.LEFT)

        self.run_status_lbl = ttk.Label(row, text="")
        self.run_status_lbl.pack(side=tk.RIGHT, padx=8)

    def set_state(self, state: RunState, *, can_start: bool) -> None:
        is_running = state is RunState.RUNNING
        self.run_btn.configure(
            state=tk.NORMAL if can_start else tk.DISABLED, text="Running" if is_running else "Start"
        )
        self.stop_btn.configure(state=tk.NORMAL if is_running else tk.DISABLED)
        self.run_status_lbl.configure(
            text={RunState.RUNNING: "Running", RunState.COMPLETE: "Complete"}.get(state, "")
        )

File: src/main/java/duckpond/buriedincode/badger/ui/RunPanel.java
package duckpond.buriedincode.badger.ui;

import java.awt.BorderLayout;
import java.awt.FlowLayout;
import javax.swing.BorderFactory;
import javax.swing.JButton;
import javax.swing.JLabel;
import javax.swing.JPanel;
import org.jspecify.annotations.NullMarked;

@NullMarked
public final class RunPanel extends JPanel {
  private final JButton runBtn;
  private final JButton stopBtn;
  private final JLabel statusLabel;

  public RunPanel(Runnable onRun, Runnable onStop) {
    super(new BorderLayout());
    setBorder(BorderFactory.createTitledBorder("Step 3  -  Run"));

    var row = new JPanel(new FlowLayout(FlowLayout.LEFT, 8, 6));
    runBtn = new JButton("Start");
    runBtn.addActionListener(e -> onRun.run());
    row.add(runBtn);

    stopBtn = new JButton("Stop");
    stopBtn.addActionListener(e -> onStop.run());
    row.add(stopBtn);
    add(row, BorderLayout.WEST);

    statusLabel = new JLabel("");
    var statusRow = new JPanel(new FlowLayout(FlowLayout.RIGHT, 8, 6));
    statusRow.add(statusLabel);
    add(statusRow, BorderLayout.EAST);
  }

  public void setState(RunState state, boolean canStart) {
    var isRunning = state == RunState.RUNNING;
    runBtn.setEnabled(canStart);
    runBtn.setText(isRunning ? "Running" : "Start");
    stopBtn.setEnabled(isRunning);
    statusLabel.setText(
        switch (state) {
          case RUNNING -> "Running";
          case COMPLETE -> "Complete";
          case IDLE -> "";
        }
    );
  }
}
